crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -419,6 +423,32 @@ fn extract_invalidated_order_uids(
         .collect()
 }
 
+/// Temporarily stores metrics for onchain order parsing, which will later be written to the global
+/// metrics object.
+struct MetricsOnchainOrderParsing {
+    errors: Arc<RwLock<HashMap<String, u64>>>,
+}
+impl MetricsOnchainOrderParsing {
+    fn new() -> Self {
+        Self {
+            errors: Arc::new(RwLock::new(HashMap::<String, u64>::new())),
+        }
+    }
+    fn write_to_metrics(self, metrics: &'static Metrics) {
+        let MetricsOnchainOrderParsing { errors } = self;
+        if let Ok(read_lock) = errors.read() {
+            for (error_label, increment) in read_lock.iter() {
+                metrics
+                    .onchain_order_errors
+                    .with_label_values(&[error_label])
+                    .inc_by(*increment);
+            }
+        } else {
+            tracing::warn!("Unable to write onchain order error metrics, poisoned RW lock");
```

Comment:
Usually we just unwrap those errors because they can only happen when a thread panics while holding a lock.
This can not happen during this call to `.read()` or the 2 calls to `.write()`.

Reply:
This comment doesn't apply anymore after the simplification but I'm still curious about it: the increment was handled in multiple futures, which in theory could be handled by multiple threads, one of which could panic (for example in `unreacable!`). Is this still unnecessary to handle?

Reply:
I think there was a misconception. "While holding the lock" wasn't supposed to mean "while the thread has access to the `RwLock`. It rather means the lock gets poisoned if the thread panics in the critical section:
```rust
let lock_guard = rwlock.read(); // lock was acquired; critical section begins
// if the thread panics here the rwlock gets poisoned
drop(lock_guard); // lock gets released; critical section ends
// if the thread panics here the rwlock is unaffected
```

In the earlier version of the PR you never did anything that could cause the thread to panic while in the critical section. You only incremented a counter. Technically this could panic in a debug build when the counter overflows but that would never happen.

Reply:
Generally we assume that the whole program exits if a thread panics. It is too much of a hassle to try to write code that keeps working when any thread could panic at any time. This is enforced by a panic handler that exits the process on any panic. You can always unwrap mutex lock calls because those only error when the mutex is poisoned by a thread panicing while holding the lock.